go-controller/pkg/ovn/egressgw.go

```diff
@@ -630,3 +607,226 @@ func (oc *Controller) cleanECMPRoutes() {
+					}
+					if _, ok := clusterRouteCache[podIP.String()]; ok {
+						clusterRouteCache[podIP.String()] = append(clusterRouteCache[podIP.String()], gwIP.String())
+					} else {
+						clusterRouteCache[podIP.String()] = []string{gwIP.String()}
+					}
+				}
+			}
+		}
+	}
+
+	// Get all namespaces with exgw routes specified
+	namespaces, err := oc.watchFactory.GetNamespaces()
+	if err != nil {
+		klog.Errorf("Error getting all namespaces for exgw ecmp route sync: %v", err)
+	}
+	for _, namespace := range namespaces {
+		if _, ok := namespace.Annotations[routingExternalGWsAnnotation]; !ok {
+			continue
+		}
+		// namespace has exgw routes, build cache
+		gwIPs, err := parseRoutingExternalGWAnnotation(namespace.Annotations[routingExternalGWsAnnotation])
+		if err != nil {
+			klog.Errorf("Unable to clean ExGw ECMP routes for namespace: %s, %v", namespace.Name, err)
+			continue
+		}
+		// get all pods in the namespace
+		nsPods, err := oc.watchFactory.GetPods(namespace.Name)
+		if err != nil {
+			klog.Errorf("Unable to clean ExGw ECMP routes for namespace: %s, %v",
+				namespace, err)
+			continue
+		}
+		for _, gwIP := range gwIPs {
+			for _, nsPod := range nsPods {
+				for _, podIP := range nsPod.Status.PodIPs {
+					if utilnet.IsIPv6(gwIP) != utilnet.IsIPv6String(podIP.IP) {
+						continue
+					}
+					if _, ok := clusterRouteCache[podIP.String()]; ok {
+						clusterRouteCache[podIP.String()] = append(clusterRouteCache[podIP.String()], gwIP.String())
+					} else {
+						clusterRouteCache[podIP.String()] = []string{gwIP.String()}
+					}
+				}
+			}
+		}
+	}
+
+	// compare caches and see if OVN routes are stale
+	for podIP, ovnRoutes := range ovnRouteCache {
+		// pod IP does not exist in the cluster
+		// remove route and any hybrid policy
+		if _, ok := clusterRouteCache[podIP]; !ok {
+			continue
+		}
+
+		// podIP exists, check if route matches
+		expectedNexthops := clusterRouteCache[podIP]
+		for _, ovnRoute := range ovnRoutes {
+			for _, clusterNexthop := range expectedNexthops {
+				if ovnRoute.nextHop == clusterNexthop {
+					ovnRoute.shouldExist = true
+				}
+			}
+		}
+	}
+
+	// iterate through ovn routes and remove any stale entries
+	for podIP, ovnRoutes := range ovnRouteCache {
+		podHasAnyECMPRoutes := false
+		for _, ovnRoute := range ovnRoutes {
+			if !ovnRoute.shouldExist {
+				klog.Infof("Found stale exgw ecmp route, podIP: %s, nexthop: %s, router: %s",
+					podIP, ovnRoute.nextHop, ovnRoute.router)
+				_, stderr, err = util.RunOVNNbctl("--if-exists", "remove", "Logical_Router",
+					strings.TrimSuffix(ovnRoute.router, "\n"), "static_routes", ovnRoute.uuid)
+				if err != nil {
+					klog.Errorf("Failed to destroy Logical_Router_Static_Route %s, stderr: %q, (%v)",
+						ovnRoute.uuid, stderr, err)
+				}
+			} else {
+				podHasAnyECMPRoutes = true
+			}
+		}
+
+		// if pod had no ECMP routes we need to make sure we remove any logical route policy for local gw mode
+		if !podHasAnyECMPRoutes {
```

Comment:
this should probably be 
```
if !podHasAnyECMPRoutes || config.GatewayMode == shared {
```

Since we need to also remove any of these LRPs when upgrading from local to shared gw mode.

Reply:
good catch, fixing